staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -248,81 +246,57 @@ func (c *Controller) Run(ctx context.Context, workers int, controllerManagerMetr
 	}
 
 	for i := 0; i < workers; i++ {
```

Comment:
I am confused on the role of "workers" here.

I sort of understand having N workers for service updates, since they can be processed in parallel.

But node updates are locked, so why do we need more than one goroutine for node workers?

Reply:
Then later, in `updateLoadBalancerHosts` we call `workqueue.ParallelizeUntil` with `workers` again.  So from a concurrency POV:

1 informer
  N service handlers
  N node handlers, but serialized, really 1
    N LB sync handlers

If not for that lock, we'd have N^2 ?

Reply:
> But node updates are locked, so why do we need more than one goroutine for node workers?

I was confused, this is wrong and thanks for spotting it. I am updating that now. 

> If not for that lock, we'd have N^2 ?

No. The "N LB sync handlers" you are referring to (from `updateLoadBalancerHosts` as you mentioned) are initialized from the node sync path - which, as you also correctly mentioned, only has 1 handler. So what we have is;

N service handler
1 node handler 
N LB sync handlers 

So, concretely: N handlers for services events, N handlers for LB sync spawned by that 1 node event handler. The service handler doesn't use the same code path and never ends up calling `updateLoadBalancerHosts` which spawns those N LB sync handlers.